src/burnchains/burnchain.rs

```diff
@@ -164,68 +165,107 @@ impl BurnchainStateTransition {
-            if let Some(missed_commit_in_block) = missed_commits_map.remove(&sortition_id) {
-                missed_commits_at_height
-                    .extend(missed_commit_in_block.into_iter().map(|x| x.clone()));
+            for blocks_back in 0..(MINING_COMMITMENT_WINDOW - 1) {
+                if parent_snapshot.block_height < (blocks_back as u64) {
+                    debug!("Mining commitment window shortened because block height is less than window size";
+                           "block_height" => %parent_snapshot.block_height,
+                           "window_size" => %MINING_COMMITMENT_WINDOW);
+                    break;
+                }
+                let block_height = parent_snapshot.block_height - (blocks_back as u64);
+                let sortition_id = match sort_tx.get_block_snapshot_by_height(block_height)? {
+                    Some(sn) => sn.sortition_id,
+                    None => break,
+                };
+                windowed_block_commits.push(SortitionDB::get_block_commits_by_block(
+                    sort_tx.tx(),
+                    &sortition_id,
+                )?);
+                let mut missed_commits_at_height =
+                    SortitionDB::get_missed_commits_by_intended(sort_tx.tx(), &sortition_id)?;
+                if let Some(missed_commit_in_block) = missed_commits_map.remove(&sortition_id) {
+                    missed_commits_at_height
+                        .extend(missed_commit_in_block.into_iter().map(|x| x.clone()));
+                }
+
+                windowed_missed_commits.push(missed_commits_at_height);
             }
+        } else {
+            // PoX reward-phase is not active
+            debug!(
+                "Block {} is in a prepare phase or post-PoX sunset, so no windowing will take place",
+                parent_snapshot.block_height + 1
+            );
 
-            windowed_missed_commits.push(missed_commits_at_height);
+            assert_eq!(windowed_block_commits.len(), 1);
+            assert_eq!(windowed_missed_commits.len(), 0);
         }
 
         // reverse vecs so that windows are in ascending block height order
         windowed_block_commits.reverse();
         windowed_missed_commits.reverse();
 
-        // figure out if the PoX sunset finished during the window
+        // figure out if the PoX sunset finished during the window,
+        // and/or which sortitions must be PoB due to them falling in a prepare phase.
         let window_end_height = parent_snapshot.block_height + 1;
         let window_start_height = window_end_height + 1 - (windowed_block_commits.len() as u64);
-        let sunset_finished_at = if sunset_end <= window_start_height {
-            Some(0)
+        let mut burn_blocks = vec![false; windowed_block_commits.len()];
+
+        // set burn_blocks flags to accomodate PoX sunset
+        if sunset_end <= window_start_height {
```

Comment:
Nitpick: can you change the setting of `burn_blocks` to a single pass over the `burn_blocks`?

```rust
           for (i, b) in burn_blocks.iter_mut().enumerate() {
                if sunset_end <= window_start_height + (i as u64) {
                    // must burn here
                    *b = true;
                } else if burnchain.is_in_prepare_phase(window_start_height + (i as u64)) {
                    // must burn
                   *b = true;
                } else {
                    // may not burn here
                    *b = false;
                }
           }
```

I think it would make it more clear how the burn_blocks vec is set (as opposed to the various special cases on `sunset_end` and multiple passes over the vec)

Reply:
Addressed in 20f7304